### PR TITLE
Add BuildTargets.md

### PR DESCRIPTION
**Pull request by lukewagner**
This PR adds BuildTargets.md to define this new concept of "build targets" as presented in both [CG-06](https://docs.google.com/presentation/d/1Y-R2fsvsurfT1cpKMvxERnqAPvWlLgiEESiheUj-QUY/edit#slide=id.g2e1fc4d92ea_0_223) and [WASI-06-12](https://github.com/WebAssembly/meetings/blob/main/wasi/2024/WASI-06-13.md), which itself was a revision of the earlier "`wasit2`" idea that came up in [WASI/#595](https://github.com/WebAssembly/WASI/issues/595#issuecomment-2087137863).

Currently, BuildTargets.md only defines one build target, `wasm32`, but with the intention of adding more in the future (e.g., `wasm64` and `wasmgc`).  See BuildTargets.md in this PR for more details.

The goal, mentioned at the end of BuildTargets.md, is that `wasi-sdk` and various other currently-component-producing toolchains would be able to emit core modules that matched `wasm32` as defined here (with "componentification" being merely a final linker option of whether to run `wasm-ld` or `wasm-component-ld`).  I plan to leave this open until we can get some implementation feedback to test this out.

### CODE REVIEW

**Comment by elewis787** [comment]
Would love to see this added and I am happy to test this. Let me know how/if I can support.

**Comment by lukewagner** [comment]
@elewis787 Great to hear!  Probably the next thing we need for this PR to merge is a concrete implementation just to validate what's here makes sense.  In particular, one great proof-of-concept would be extending the [wit-component crate](https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wit-component) to accept any core module targeting `wasm32` as defined here and emit a `wasip2` component that runs on unmodified wasmtime or jco.

**Comment by elewis787** [comment]
@lukewagner sounds great. 

A very basic example showing wasip1 and wasip2 and using WIT to generate bindings can be found here https://github.com/elewis787/wasip1-wasip2. 

The goal is to show that Tinygo ( dev branch supporting wasip1 and wasip2 ) can be used to target a component or a module. Once these are compiled, this example shows that the wasmtime-go bindings ( currently wasip1 support only ) can be used for core modules, and wasmtime-cli can be used for components. 

Let me know if this is on the right track for what you would like to see. 

Regardless, I will start digging into the wit-component crate.

**Comment by jeff1010322** [comment]
I am looking into this as well. Does this mean there would be build target support for generating modules (wasip1 only) versus generating components (wasip2) from compilers?

I have been testing out WebAssembly modules with JCO and through their `transpile` command I can convert the component into a core module that can be run in runtimes that don't support components like wasmtime-go. I just have to make sure the runtime links any imports called by the core module. In JCO you can manually disable the WASI P2 features with a few flags.

But is there a plan to make core modules without WASI P2 imports a build target for things like this?

**Comment by alexcrichton** [comment]
From a technical perspective almost all components today start as single core modules. This is the case for nearly all compiles for TinyGo, Rust, and C (e.g. LLVM-based toolchains). In all of these cases LLVM emits a single core wasm module which is then fed into `wit-component` which generates a component. So in that sense the answer to:

> But is there a plan to make core modules without WASI P2 imports a build target for things like this?

is "yes that's sort of already supported today". I say "sort of" in that the way this typically works is that the original binary is using a mixture of `wasi_snapshot_preview1` APIs plus WIT-bound APIs. That means that it's sort of a hybrid p1/p2 module which doesn't have necessarily clear semantics. The componentization process papers over all of this to produce a single component out the other end.

The goal of this proposal is to formalize the core-module-to-component step of tooling. It makes the core module artifact (which today becomes a component) a first-class artifact. It additionally defines what it means to import/export various names from a core module and this proposal means that most core modules emitted by compilers today are **not valid to become a component** because they import `wasi_snapshot_preview1. 

This means that one big chunk in implementing this proposal is going to be updating language standard libraries to exclusively use WASIp2 APIs instead of a mixture of WASIp1 and WASIp2. This would likely start with wasi-libc and start bubbling up from there for example.

**Comment by elewis787** [comment]
Thanks @alexcrichton! This is helpful and clear. 

as a naive example, does this demonstrate what you are discussing? 

`import "wasi_snapshot_preview1" "fd_write"` with  wasip1 
vs 
`import "wasi:filesystem/types@0.2.0" "[method]descriptor.write"` with wasip2

If so, is a byproduct of this the core-module that is produced with wasip2 imports ? If so, in the example I provided the obvious difference is that the module produced does not use wasip2 syntax. 

Are there any plans for taking a component to a module to be backwards compatible with wasip1? 

Thanks for walking through regardless!

**Comment by alexcrichton** [comment]
Indeed yeah the differences will come down to imports. The first one you listed is WASIp1, and the second one is WASIp2-as-defined-by-`wit-component`-today which is not a standard and does not match what this document is proposing. This document would use a WASIp2 target that looks like:

```wasm
(import "cm32p2|ns:wasi:filesystem/types@0.2" "[method]descriptor.write" (func (param ...)))
```

Notably the prefix `cm32p2|ns:` on the import moudle plus the lack of trailing `.0` in the import module as well.

> If so, is a byproduct of this a core module that is produced but using wasip2 imports before componentization?

Yes, and this document is describing the exact name of component imports/exports to be "componentized", optionally, later on. This document is a formalization of the names used today which is intended to become standard (as opposed to whatever was implemented to begin with)

> Are there any plans for taking a component to a module to be backward compatible with wasip1? 

Not currently, but that's also sort of out of the scope of this PR

**Comment by elewis787** [comment]
@alexcrichton, thanks for explaining this. 

@jeff1010322 and I are both in the jco exploration phase. As mentioned, we are playing around with the transpile flags to produce core modules. As a next step, do you feel it is helpful to look into the generation of the core module and take a stab at outlining the behavior listed here? 

As far as backward compatibility, do you know of a place to discuss/contribute to this? 

After using WIT and a few various tools, it seems like there may be some options/benefits in the bindings generated. Candidly, I am still getting up to speed with WASIp2 and may be incorrect.

**Comment by alexcrichton** [comment]
Oh no worries! Always great to have more folks help and happy to help out where I can!

I think the first step is what you're already doing here which is to make sure you feel comfortable with the current process of how a component is created. The high-level overview of that is that source code is written with the help of `wit-bindgen` frequently, compiled with a language's standard library, and assembled with LLVM's `wasm-ld` to produce a core wasm module. This module is fed through `wit-component` to produce a component. 

The next step is going to be working backwards in this pipeline from the end back to the front. For example if you were to change the core wasm module today nothing would be able to turn it into a component. In that sense you'll want to do as @lukewagner suggested which is to start with `wit-component`, the final layer of this stack. That lives in the wasm-tools repository and you'll basically be updating various locations of where a core wasm "thing" is matched up to a component model "thing" to match the conventions outlined in this document. That repository has a whole bunch of handwritten tests in `*.wat` syntax which you can add to for this.

Once that's all implemented the next step is to probably work on `wit-bindgen`. That'll be updating all of its generated code to generate references to new names for imports/exports. That'll probably feel more natural once `wit-component` implemented.

From there the next hard part will be to flesh out language standard libraries, but that's probably best tackled once we're closer to that.

**Comment by elewis787** [comment]
Great overview. Thanks again! 

I'll start looking into this. I am still getting up to date on the various tools but I believe I understand what's required now.

**Comment by alexcrichton** [comment]
I have an implementation of this at https://github.com/bytecodealliance/wasm-tools/pull/1828 which I believe captures everything here.

**Comment by lukewagner** [comment]
Nice!  Let me know once things seem far enough along that you feel confident that what's in this PR can match reality.

**Review by sunfishcode — COMMENTED** [review]
Overall this looks good!

File: design/mvp/BuildTargets.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+* It allows existing *Core WebAssembly producer toolchains* to more easily
+  and incrementally target the Component Model.
+* It allows existing *Core WebAssembly runtimes* to implement WIT-defined
+  interfaces without having to implement the full Component Model, providing an
+  incremental implementation step towards the full Component Model.
+
+Furthermore, any module matching a Core WebAssembly build target can be
+trivially wrapped (e.g., by [`wasm-tools component new`]) to become a
+semantically-equivalent component and thus these modules can be considered
+**simple components**.
```
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/BuildTargets.md:42]: Maybe the wording is unclear, but the intention of this document isn't to bind to any specific world; custom worlds should be given a core module build target by this document just as well as standard ones.  There is some wording about "fixing" a world, but that's just for the purpose of the rest of the description so we can just say "the set of imports" (implicitly referring to that fixed world) instead of something more verbose.

File: design/mvp/BuildTargets.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+ 1. [WIT-derived function imports](#wit-derived-imports)
+
+and the union of the following sets of exports:
+ 1. [WIT-derived function exports](#wit-derived-exports)
+ 2. [Memory exports](#memory-exports)
+ 3. [Initialization export](#initialization-export)
+
+Every Core WebAssembly import and export name defined by the build target
+starts with a `prefix` string that incorporates the build target:
+* For `wasm32` this `prefix` string is `cm32p2`.
```
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/BuildTargets.md:96]: This hasn't merged yet (and is not, afaik, being used as part of anyone's serious workflow) so, depending on the timing of these two concurrent tasks, we could hypothetically skip right to `cm32p3`.

File: design/mvp/BuildTargets.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+build step (which aligns well with the overall [linking] toolchain pipeline).
+In this case, the toolchain unconditionally takes the Core WebAssembly build
+target as an argument and then takes an independent flag specifying whether to
+emit a Core WebAssembly module or a component.
+
+For example:
+* For `wasi-libc`-based toolchains like `wasi-sdk` or `rustc`:
+  * The `--target` is `wasm32-wasip2`, combining the `wasm32` build target
+    defined in this document with the additional information that the language
+    runtime can import [WASI Preview 2]-defined interfaces.
```
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/BuildTargets.md:412]: Agreed in large part; indeed some parts of this doc should be updated to reflect newer discussions with various language toolchains on what to do about WASI interfaces (esp. those in `wasi:cli/imports`) that are *optionally* used by various parts of the standard library and language runtime that don't *a priori* know the "target world" (which, to your point, is indeed orthogonal to the "build target").

Based on the newer discussions, I agree with your second paragraph that `cm32p2` (and/or `cm32p3`, depending on timing) are the "build target(s)" defined by this doc.  And then what this "build target" would say is:
* "you're using core wasm with 32-bit linear memory"
* "you're able to use the "Preview {2,3}" WIT/C-M feature set with the `canonopt`s set by this document"
* "you're aware of the *existence* of the WASI interfaces that have reached Phase 3 and have been stamped with a 0.{2,3}.* version, but whether they are implemented by the host runtime is not guaranteed by the build target -- the interfaces may or may not be implemented by the host and thus you should *optionally import* them if you don't know for sure that the target world (which is an orthogonal concept from 'build target') includes them".

This leaves open the question of how one "optionally imports".  In the medium-term the plan is to add `optional` to the C-M/WIT, but in the short-term in BuildTargets.md we could simply emit 2 core function imports for each optional import (the one to ask whether you got it and the one that traps if you call it and you didn't get it) with appropriately-mangled names.

Then in, e.g., wasi-sdk, we could say that `--target=wasm32-wasip2` is the CLI syntax to ask for the `cm32p2` build target whereas in Go, `GOARCH=wasm32` and `GOOS=wasip2` is the buildconfig to ask for the same `cm32p2` buildtarget.  The target-world-agnostic lang runtime and stdlib code would then optionally import everything it might want (with graceful fallback if the "did I get it?" function returns "no"), and then `wasm-component-ld` (which *does* know the target world) can trivially implement the core function imports appropriately based on this knowledge.

Does that make sense?

File: design/mvp/BuildTargets.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (sunfishcode) [reply design/mvp/BuildTargets.md:42]: Ah, that wasn't clear to me.

I'd suggest rewording the sentence "The rest of this document assumes a single, fixed "target world" to say something more like "build targets are paired with a target world, producing a complete ABI".

Another particular thing that confused me was that the document says build targets provide that "any module matching a Core WebAssembly build target can be trivially wrapped", but this is only true if `component new` knows the world, or the module contains the special custom section produced by wit-bindgen, which I guessed you wouldn't want to depend on here.

File: design/mvp/BuildTargets.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunfishcode) [reply design/mvp/BuildTargets.md:412]: To my ears, the name "build target" for this concept sounds like something that end users will recognize, but will recognize as something other than what you're using it to mean here.

That aside, I do like using the name "cm32p2" for this concept because that means it directly corresponds to the name prefix it uses, which is nice.

The wording "you're aware of the existence of the WASI interfaces that have [...]" could perhaps be expressed more generically, to avoid making it sound specific to WASI. Perhaps "there may be features in the target world that not all hosts provide".

File: design/mvp/BuildTargets.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/BuildTargets.md:42]: Great points!  I'll improve those sentences to be more clear.

File: design/mvp/BuildTargets.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukewagner) [reply design/mvp/BuildTargets.md:412]: I'm not wed to the phrase "Build targets" so we could change it to "Toolchain ABI", as you suggested.  (Originally, I had thought that there would be both "core module" and "component" "build targets", but that has since shifted so that now we're only talking about core modules (that may or may not get wrapped into components), so "ABI" makes a lot more sense now.)

> The wording "you're aware of the existence of the WASI interfaces that have [...]" could perhaps be expressed more generically, to avoid making it sound specific to WASI. Perhaps "there may be features in the target world that not all hosts provide".

That's a fair point.  But `cm32p2` *is* specific to the amorphous "Preview {2,3}" concept.  Previews are meant to go away when we hit 1.0, of course, leaving presumably `cm32` as the Build Target / Toolchain ABI.  And then yeah, whether you choose to (optionally or not) import WASI is orthogonal to the use of `cm32` (or `cm64` or `cm-gc`).

**Comment by bashor** [comment]
@lukewagner, I'm wondering what the status of the PR is and if you're planning to merge it.

**Comment by lukewagner** [comment]
@bashor Good question!  IIUC, this got partially-but-not-100% implemented in `wasm-component-ld` and other parts of the toolchain and then a lot of our focus shifted to finishing 0.3 and that's why this PR hasn't merged yet.  With 0.3 very-almost done, it's now a question of whether we should skip `cm32p2` and do a `cm32p3` instead.  There's also a tentative plan to work on #383 in the short-term, which would have the effect of removing `cm32p3_realloc` and changing the ABI, so there's another question of whether to base `cm32p3` on that.  If we had infinite resources, I think we'd finish off `cm32p2` ASAP and then release a `cm32p3` after the dust settled on #383, but since we don't, I don't know; happy to hear what other folks think.

**Comment by jeffparsons** [comment]
> There's also a tentative plan to work on #383 in the short-term

Ooooh, that's exciting! 👀

> happy to hear what other folks think.

I personally don't see much point in spending too much work fleshing out additional modes of operation for p2; it's already an "island of stability", as they say, and with p3 around the corner I'd imagine most folks looking for the new hotness would be willing to track that instead. (E.g. for my hobby projects I'm targeting the wasip3 drafts...)

But, y'know, sample size = 1.
